beginner_source/translation_transformer.py

```diff
@@ -8,187 +8,154 @@
-                            dtype=torch.long)
-    data.append((de_tensor_, en_tensor_))
-  return data
 
+vocab_transform = {}
 
-train_data = data_process(train_filepaths)
-val_data = data_process(val_filepaths)
-test_data = data_process(test_filepaths)
-device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+# Build source vocabulary
+train_iter = Multi30k(split='train', language_pair=(SRC_LANGUAGE, TGT_LANGUAGE))
+vocab_transform[SRC_LANGUAGE] = build_vocab_from_iterator(yield_tokens(train_iter, SRC_LANGUAGE))
 
+# Build target vocabulary
+train_iter = Multi30k(split='train', language_pair=(SRC_LANGUAGE, TGT_LANGUAGE))
+vocab_transform[TGT_LANGUAGE] = build_vocab_from_iterator(yield_tokens(train_iter, TGT_LANGUAGE))
 
-BATCH_SIZE = 128
-PAD_IDX = de_vocab['<pad>']
-BOS_IDX = de_vocab['<bos>']
-EOS_IDX = de_vocab['<eos>']
-
-
-######################################################################
-# DataLoader
-# ----------
-# 
-# The last torch specific feature we’ll use is the DataLoader, which is
-# easy to use since it takes the data as its first argument. Specifically,
-# as the docs say: DataLoader combines a dataset and a sampler, and
-# provides an iterable over the given dataset. The DataLoader supports
-# both map-style and iterable-style datasets with single- or multi-process
-# loading, customizing loading order and optional automatic batching
-# (collation) and memory pinning.
-# 
-# Please pay attention to collate_fn (optional) that merges a list of
-# samples to form a mini-batch of Tensor(s). Used when using batched
-# loading from a map-style dataset.
-# 
-
-from torch.nn.utils.rnn import pad_sequence
-from torch.utils.data import DataLoader
-
-def generate_batch(data_batch):
-  de_batch, en_batch = [], []
-  for (de_item, en_item) in data_batch:
-    de_batch.append(torch.cat([torch.tensor([BOS_IDX]), de_item, torch.tensor([EOS_IDX])], dim=0))
-    en_batch.append(torch.cat([torch.tensor([BOS_IDX]), en_item, torch.tensor([EOS_IDX])], dim=0))
-  de_batch = pad_sequence(de_batch, padding_value=PAD_IDX)
-  en_batch = pad_sequence(en_batch, padding_value=PAD_IDX)
-  return de_batch, en_batch
+# Define special symbols and indices
+UNK_IDX, PAD_IDX, BOS_IDX, EOS_IDX = 0, 1, 2, 3
+special_symbols = {'<unk>': UNK_IDX, '<pad>': PAD_IDX,'<bos>': BOS_IDX, '<eos>': EOS_IDX}
 
-train_iter = DataLoader(train_data, batch_size=BATCH_SIZE,
-                        shuffle=True, collate_fn=generate_batch)
-valid_iter = DataLoader(val_data, batch_size=BATCH_SIZE,
-                        shuffle=True, collate_fn=generate_batch)
-test_iter = DataLoader(test_data, batch_size=BATCH_SIZE,
-                       shuffle=True, collate_fn=generate_batch)
+# Insert special symbols
+for symbol, index in special_symbols.items():
+    vocab_transform[SRC_LANGUAGE].insert_token(symbol, index)
+    vocab_transform[TGT_LANGUAGE].insert_token(symbol, index)
 
+# Set default index to return OOV token is queried
```

Comment:
nit: This reads a bit weird in my mind. Maybe you mean something like "Return UNK_IDX if a token cannot be found"?

Reply:
Similar to above we could also mention that the Vocab would throw an error if this method wasn't set and explain how to unset it again and check whether it's been set.

Reply:
Sounds good! Yes, got to work on text and comments a bit :). I plan to include more elaborate description on the behavior in the text part.

Reply:
Ideally something like that would be covered in a central tutorial similar to torchaudio's [data augmentation tutorials](https://pytorch.org/tutorials/beginner/audio_preprocessing_tutorial.html#data-augmentation).

Reply:
> Ideally something like that would be covered in a central tutorial similar to torchaudio's [data augmentation tutorials](https://pytorch.org/tutorials/beginner/audio_preprocessing_tutorial.html#data-augmentation).

Yes, this seems to be a neat idea. We should plan to create tutorial on torchtext's  core utility and functionalities!